File: src/main/java/net/katsstuff/danmakucore/impl/shape/ShapeRandomRing.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.impl.shape;

import java.util.HashSet;
import java.util.Random;

import net.katsstuff.danmakucore.data.Quat;
import net.katsstuff.danmakucore.data.Vector3;
import net.katsstuff.danmakucore.entity.danmaku.DanmakuTemplate;
import net.katsstuff.danmakucore.entity.danmaku.EntityDanmaku;
import net.katsstuff.danmakucore.shape.IShape;
import net.katsstuff.danmakucore.shape.ShapeResult;

public class ShapeRandomRing implements IShape {

	private final DanmakuTemplate danmaku;
	private final int amount;
	private final float radius;
	private final double distance;

	public ShapeRandomRing(DanmakuTemplate danmaku, int amount, float radius, double distance) {
		this.danmaku = danmaku;
		this.amount = amount;
		this.radius = radius;
		this.distance = distance;
	}

	@Override
	public ShapeResult draw(Vector3 pos, Quat orientation, int tick) {
		HashSet<EntityDanmaku> set = new HashSet<>();
		if(!danmaku.world.isRemote) {
			Random rand = danmaku.world.rand;
			Quat rotatedOrientation = orientation.multiply(Quat.fromAxisAngle(Vector3.Right(), 90));

			for(int i = 0; i < amount; i++) {
				Quat rotate = rotatedOrientation.multiply(Quat.fromAxisAngle(Vector3.Up(), rand.nextDouble() * 360D)
						.multiply(Quat.fromAxisAngle(Vector3.Left(), 90D - rand.nextDouble() * radius)));
				danmaku.direction = Vector3.Forward().rotate(rotate);
				danmaku.pos = pos.offset(danmaku.direction, distance);
				EntityDanmaku spawned = this.danmaku.asEntity();
				danmaku.asEntity().world.spawnEntity(spawned);
				set.add(spawned);
			}
		}
		return ShapeResult.done(set);
	}
}

File: src/main/scala/net/katsstuff/danmakucore/impl/shape/ShapeRandomRing.scala
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.impl.shape

import net.katsstuff.danmakucore.danmaku.{DanmakuState, DanmakuTemplate}
import net.katsstuff.mirror.data.{Quat, Vector3}
import net.katsstuff.danmakucore.shape.{Shape, ShapeResult}

class ShapeRandomRing(template: DanmakuTemplate, amount: Int, radius: Float, distance: Double) extends Shape {
  override def draw(pos: Vector3, orientation: Quat, tick: Int): ShapeResult = {
    if (!template.world.isRemote) {
      val rand               = template.world.rand
      val rotatedOrientation = orientation.multiply(Quat.fromAxisAngle(Vector3.Right, 90))
      val rotate = rotatedOrientation.multiply(
        Quat
          .fromAxisAngle(Vector3.Up, rand.nextDouble * 360D)
          .multiply(Quat.fromAxisAngle(Vector3.Left, 90D - rand.nextDouble * radius))
      )

      val builder = template.toBuilder
      val res = for (_ <- 0 until amount) yield {
        builder.direction = Vector3.Forward.rotate(rotate)
        builder.pos = pos.offset(builder.direction, distance)
        builder.build.asEntity
      }

      ShapeResult.done(res.toSet)
    } else ShapeResult.done(Set.empty[DanmakuState])
  }
}
